File: version4j/src/main/java/com/only4/algorithm/version4j/leetcode/CopyRandomList.java
package com.only4.algorithm.version4j.leetcode;

import com.only4.algorithm.version4j.extra.Node;

public class CopyRandomList {
    public Node copyRandomList(Node head) {
        if (head == null) return null;

        createCopyNodes(head);
        assignRandomPointers(head);
        return separateLists(head);
    }

    private void createCopyNodes(Node head) {
        Node current = head;
        while (current != null) {
            Node copyNode = new Node(current.val);
            copyNode.next = current.next;
            current.next = copyNode;
            current = copyNode.next;
        }
    }

    private void assignRandomPointers(Node head) {
        Node current = head;
        while (current != null) {
            if (current.random != null) {
                current.next.random = current.random.next;
            }
            current = current.next.next;
        }
    }

    private Node separateLists(Node head) {
        Node copyHead = head.next;
        Node current = head;
        while (current != null && current.next != null) {
            Node temp = current.next;
            current.next = current.next.next;
            current = temp;
        }
        return copyHead;
    }
}

File: version4k/src/main/kotlin/com/only4/algorithm/version4k/leetcode/copyRandomList.kt
package com.only4.algorithm.version4k.leetcode

/**
 * 链表节点定义，包含值、指向下一个节点的指针和随机指针
 */
class Node(var `val`: Int) {
    var next: Node? = null
    var random: Node? = null
}

fun copyRandomList(head: Node?): Node? {
    head ?: return null

    fun createCopyNodes() {
        var current: Node? = head
        while (current != null) {
            val copyNode = Node(current.`val`)
            copyNode.next = current.next
            current.next = copyNode
            current = copyNode.next
        }
    }

    fun assignRandomPointers() {
        var current: Node? = head
        while (current != null) {
            current.next?.random = current.random?.next
            current = current.next?.next
        }
    }

    fun separateLists(): Node? {
        val copyHead = head.next
        var current: Node? = head
        while (current?.next != null) {
            val temp = current.next
            current.next = current.next?.next
            current = temp
        }
        return copyHead
    }

    createCopyNodes()
    assignRandomPointers()
    return separateLists()
}
